solidity/contracts/vault/TBTCOptimisticMinting.sol

```diff
@@ -0,0 +1,219 @@
+// SPDX-License-Identifier: GPL-3.0-only
+
+// ██████████████     ▐████▌     ██████████████
+// ██████████████     ▐████▌     ██████████████
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+// ██████████████     ▐████▌     ██████████████
+// ██████████████     ▐████▌     ██████████████
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+
+pragma solidity 0.8.17;
+
+import "../bridge/Bridge.sol";
+import "../bridge/Deposit.sol";
+
+abstract contract TBTCOptimisticMinting is Ownable {
+    // TODO: make it governable?
+    uint256 public constant OPTIMISTIC_MINTING_DELAY = 3 hours;
+
+    Bridge public bridge;
+
+    mapping(address => bool) public isMinter;
+    mapping(address => bool) public isGuard;
+
+    mapping(uint256 => uint256) public pendingOptimisticMints;
+
+    mapping(address => uint256) public optimisticMintingDebt;
+
+    event OptimisticMintingRequested(
+        address indexed minter,
+        bytes32 fundingTxHash,
+        uint32 fundingOutputIndex,
+        uint256 depositKey
+    );
+    event OptimisticMintingFinalized(
+        address indexed minter,
+        bytes32 fundingTxHash,
+        uint32 fundingOutputIndex,
+        uint256 depositKey
+    );
+    event OptimisticMintingCancelled(
+        address indexed guard,
+        bytes32 fundingTxHash,
+        uint32 fundingOutputIndex,
+        uint256 depositKey
+    );
+    event MinterAdded(address indexed minter);
+    event MinterRemoved(address indexed minter);
+    event GuardAdded(address indexed guard);
+    event GuardRemoved(address indexed guard);
+
+    modifier onlyMinter() {
+        require(isMinter[msg.sender], "Caller is not a minter");
+        _;
+    }
+
+    modifier onlyGuard() {
+        require(isGuard[msg.sender], "Caller is not a guard");
+        _;
+    }
+
+    constructor(Bridge _bridge) {
+        require(
+            address(_bridge) != address(0),
+            "Bridge can not be the zero address"
+        );
+
+        bridge = _bridge;
+    }
+
+    function _mint(address minter, uint256 amount) internal virtual;
+
+    function optimisticMint(bytes32 fundingTxHash, uint32 fundingOutputIndex)
+        external
+        onlyMinter
+    {
+        uint256 depositKey = calculateDepositKey(
+            fundingTxHash,
+            fundingOutputIndex
+        );
+        Deposit.DepositRequest memory deposit = bridge.deposits(depositKey);
+
+        // TODO: validate when it was revealed?
+
+        require(deposit.revealedAt != 0, "The deposit has not been revealed");
+        require(deposit.sweptAt == 0, "The deposit is already swept");
+        require(deposit.vault == address(this), "Unexpected vault address");
```

Comment:
Perhaps a more intuitive error message here?  "Vault address cannot be bridge address.  Please copy and paste more carefully.  :-)"

Reply:
This address is not something the user is entering or copy-pasting. The vault's address is set during the deposit reveal, and only deposits targeted to the `TBTCVault` can be requested for optimistic minting. 

> "Vault address cannot be bridge address. Please copy and paste more carefully. :-)"

The user is not copy-pasting the value for this call. If they made a mistake during the deposit reveal, they must wait for the P2SH unlock. 

> "The deposit's vault doesn't match the current vault"

There is only one `TBTCVault`, and "current" is not changing.

Is the current revert message unintuitive? 🤔